### PR TITLE
GTEST: add tag matching tests

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1153/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3083/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1154/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3084/ for details (Mellanox internal link).

File: test/gtest/uct/test_hw_tag_matching.cc [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    typedef ucs_status_t (test_hw_tag_matching::*send_func)(entity&, send_ctx&);
+
+    void init() {
+        uct_test::init();
+
+        check_caps(TAG_CAPS);
+
+        uct_iface_params params;
+
+        // tl and dev names are taken from resources via GetParam, no need
```
Comment (shamisp) [inline test/gtest/uct/test_hw_tag_matching.cc]: Non C++ comments

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1155/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3188/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1244/ for details.

**Comment by yosefe** [comment]
@MattBBaker can you pleas post ORNL failure?

**Comment by MattBBaker** [comment]
Some C++ black magic exploded during compile.

```
make[2]: Entering directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/test/gtest'
  CXX    common/gtest-main.o
[Cut out a lot of boring lines here]
  CXX    uct/gtest-test_tag.o
cc1plus: warnings being treated as errors
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc: In static member function ‘static void test_tag::tag_consumed(uct_tag_context_t*)’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:250: error: invalid access to non-static data member ‘test_tag::recv_ctx::uct_ctx’  of NULL object
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:250: error: (perhaps the ‘offsetof’ macro was used incorrectly)
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc: In static member function ‘static void test_tag::completed(uct_tag_context_t*, uct_tag_t, uint64_t, size_t, ucs_status_t)’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:257: error: invalid access to non-static data member ‘test_tag::recv_ctx::uct_ctx’  of NULL object
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:257: error: (perhaps the ‘offsetof’ macro was used incorrectly)
In file included from /var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/test_helpers.h:11,
                 from /var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/test.h:10,
                 from /var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:10:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/gtest.h: In function ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = long unsigned int, T2 = long int]’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/gtest.h:18899:   instantiated from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = size_t, T2 = ssize_t, bool lhs_is_null_literal = false]’
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/test_tag.cc:402:   instantiated from here
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/common/gtest.h:18862: error: comparison between signed and unsigned integer expressions
make[2]: *** [uct/gtest-test_tag.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests'
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3191/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1271/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3217/ for details (Mellanox internal link).

File: test/gtest/uct/uct_test.cc [hunk 2]
```diff
@@ -156,6 +156,12 @@ uct_test::entity* uct_test::create_entity(size_t rx_headroom) {
     return new_ent;
 }
 
+uct_test::entity* uct_test::create_entity(uct_iface_params_t &params) {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: should be const uct_iface_params_t& params
also, the old create_entity(rx_headroom) should be changed to call this function

File: test/gtest/uct/test_tag.cc [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+
+extern "C" {
+#include <uct/api/uct.h>
+}
+#include <common/test.h>
+#include "uct_test.h"
+
+
+class test_tag : public uct_test
+{
```
Comment (yosefe) [inline test/gtest/uct/test_tag.cc]: code style - bracket on previous line

File: test/gtest/uct/test_tag.cc [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        params.eager_cb    = unexp_eager;
+        params.eager_arg   = NULL;
+        params.rndv_cb     = unexp_rndv;
+        params.rndv_arg    = NULL;
+
+        if (UCT_DEVICE_TYPE_SELF == GetParam()->dev_type) {
+            entity *e = uct_test::create_entity(params);
+            m_entities.push_back(e);
+
+            e->connect(0, *e, 0);
```
Comment (yosefe) [inline test/gtest/uct/test_tag.cc]: m_sender is not initialized?

File: test/gtest/uct/test_tag.cc [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+                       uct_tag_t m = MASK) {
+        r.mbuf                    = b;
+        r.tag                     = t;
+        r.tmask                   = m;
+        r.uct_ctx.completed_cb    = completed;
+        r.uct_ctx.tag_consumed_cb = tag_consumed;
+        r.comp = r.unexp = r.consumed = false;
+    }
+
+    void wait_completion (bool *event) {
```
Comment (yosefe) [inline test/gtest/uct/test_tag.cc]: can you reuse wait_for_flag?

File: test/gtest/uct/test_tag.cc [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+        uct_tag_t tag = 11;
+
+        mapped_buffer recvbuf(length, RECV_SEED, receiver());
+        mapped_buffer sendbuf(length, SEND_SEED, sender());
+        recv_ctx r_ctx;
+        init_recv_ctx(r_ctx, &recvbuf, tag);
+        send_ctx s_ctx(&sendbuf, tag, reinterpret_cast<uint64_t>(&r_ctx));
+        ASSERT_UCS_OK((this->*sfunc)(sender(), s_ctx));
+
+        short_progress_loop();
```
Comment (yosefe) [inline test/gtest/uct/test_tag.cc]: why is this needed?

File: test/gtest/uct/test_tag.cc [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+{
+  check_caps(UCT_IFACE_FLAG_TAG_EAGER_BCOPY);
+
+  uct_iface_set_am_handler(receiver().iface(), 0, am_handler,
+                           NULL, UCT_AM_CB_FLAG_SYNC);
+  mapped_buffer lbuf(200, SEND_SEED, sender());
+  ssize_t len = uct_ep_am_bcopy(sender().ep(0), 0, mapped_buffer::pack,
+                                reinterpret_cast<void*>(&lbuf));
+  EXPECT_EQ(lbuf.length(), static_cast<size_t>(len));
+  short_progress_loop();
```
Comment (yosefe) [inline test/gtest/uct/test_tag.cc]: shall we use wait_for_flag?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1288/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3233/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1289/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3234/ for details (Mellanox internal link).
